[jit] make linear avaiable in tracing

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20284 make linear avaiable in tracing**
* #20039 [jit] dispatch and expose linear op
* #19988 [jit] split canonicalize_ops, make a decompose pass

Summary:

tracing only record the c++ function we called (addmm/matmul), we should expose linear in tracing since we did the decomposition pass to decompose linear

wanchaol committed:
make linear avaiable in tracing

wanchaol committed:
Update on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

wanchaol committed:
add decompose pass to onnx on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

wanchaol committed:
small AD optimize on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

wanchaol committed:
add onnx symbolic for linear on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

wanchaol committed:
add one more test case on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

bddppq commented on torch/onnx/symbolic.py:
```diff
@@ -355,6 +355,16 @@ def addmm(g, self, mat1, mat2, beta, alpha):
 
+def linear(g, input, weight, bias):
+    if input.type().kind() in ["CompleteTensorType", "DimensionedTensorType"] and input.type.dim() == 2 and not bias.node().mustBeNone():
+        return addmm(g, bias, input, t(g, weight))
```
This will materialize the transpose, onnx Gemm has transA and transB attributes

wanchaol committed:
rebase with master on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

wanchaol committed:
Update on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

wanchaol committed:
Update on "[jit] make linear avaiable in tracing"

make linear avaiable in tracing

gh-metadata: pytorch pytorch 20284 gh/wanchaol/5/head

garymm commented:
Closing as obsolete.
It seems like ONNX exporter now sees aten::linear, which makes me think it got implemented at some point?
See #54866
